# 1. Introduction
1.1. Name: Nurvita Aisah

1.2. Batch: FTDS 007

# 2.Problem
On search engine algorithm, Google implements Page Rank algorithm to determine the most important web page based on search keywords. As a result, Google is able to see how many connections are between web pages. This algorithm is frequently used by Social Scientists for Social Network Analysis.

  For instance, based on a search keyword, a web page group has a web connection as follows:
<p align="center"><img src="https://www.intmath.com/matrices-determinants/svg/svgphp-applications-eigenvalues-eigenvectors-8-s0.svg"></img></p>

  Based on a web page group above, which is the most important webpage?



# 3. Libraries

Used libraries

In [21]:
import numpy as np
from numpy.linalg import eig

# 4. Solution  
  **Hint**:
  - To solve the problem, use Eigendecomposition concept.
  - The most important web page is the web page that has the highest eigenvalue.
  - To determine Page Rank value, use an eigenvector which has $\lambda$ positive and real. (Sometime <img src="https://latex.codecogs.com/png.latex?\lambda=1" title="\lambda=1" /> is applied for Page Rank case)
  - Create a matrix <img src="https://latex.codecogs.com/png.latex?\mathbf{A}" title="\mathbf{A}" />, as follows:
 
    For instance, we use web page 1’s case. At web page 1, there are 4 out connections from web page 1 (to web page 2, 4, 5, 6). Hence, the value of matrix’s elements are Sehingga nilai elemen matriks <img src="https://latex.codecogs.com/png.latex?A[1,2]=A[1,4]=A[1,5]=A[1,6]=\frac{1}{4}" title="A[1,2]=A[1,4]=A[1,5]=A[1,6]=\frac{1}{4}" />,  the element which has no web page connection will be valued as 0. 



**Breakdown the web page into a matrix**

Page 1 = A[1,2]= A[1,4]= A[1,5]=A[1,6]= 1/4

Page 2 = B[2,3]= B[2,4]= 1/2

Page 3 = C[3,5]= 1
 
Page 4 = D[4,5]= 1

Page 5 = E[5,1]= E[5,4]= 1/2

Page 6 = F[6,5]= 1

**Matrix Page Rank**

In [19]:
a = [[0,0,0,0,0.5,0],[0.25,0,0,0,0,0],[0,0.5,0,0,0,0],[0.25,0.5,0,0,0.5,0],[0.25,0,1,1,0,1],[0.25,0,0,0,0,0]]
A = np.array(a)
A

array([[0.  , 0.  , 0.  , 0.  , 0.5 , 0.  ],
       [0.25, 0.  , 0.  , 0.  , 0.  , 0.  ],
       [0.  , 0.5 , 0.  , 0.  , 0.  , 0.  ],
       [0.25, 0.5 , 0.  , 0.  , 0.5 , 0.  ],
       [0.25, 0.  , 1.  , 1.  , 0.  , 1.  ],
       [0.25, 0.  , 0.  , 0.  , 0.  , 0.  ]])

**Seach the eigenvalue and eigenvector of web page**

In [23]:
eigenvalue, eigenvector = eig(A)
print("Eigenvalue : ", eigenvalue)
print("Eigenvector: ", eigenvector)

Eigenvalue :  [ 1.00000000e+00+0.j         -1.39845075e-01+0.39240258j
 -1.39845075e-01-0.39240258j -7.20309850e-01+0.j
  6.27089983e-17+0.j         -2.02451367e-17+0.j        ]
Eigenvector:  [[ 3.77123617e-01+0.j         -5.77040252e-01+0.j
  -5.77040252e-01-0.j          4.89508017e-01+0.j
  -2.95787278e-16+0.j          3.83213810e-16+0.j        ]
 [ 9.42809042e-02+0.j          1.16252589e-01+0.32620252j
   1.16252589e-01-0.32620252j -1.69894948e-01+0.j
  -1.72542579e-16+0.j          0.00000000e+00+0.j        ]
 [ 4.71404521e-02+0.j          3.21965131e-01-0.2628717j
   3.21965131e-01+0.2628717j   1.17931851e-01+0.j
  -7.54422638e-01+0.j         -3.73042501e-03+0.j        ]
 [ 5.18544973e-01+0.j         -1.38822532e-01+0.06333082j
  -1.38822532e-01-0.06333082j  4.37544920e-01+0.j
   6.47645899e-01+0.j         -7.05234189e-01+0.j        ]
 [ 7.54247233e-01+0.j          1.61392475e-01-0.45286417j
   1.61392475e-01+0.45286417j -7.05194893e-01+0.j
   2.23498431e-16+0.j         -6.90384823

# 5. Conclusion
Based on eigenvector and eigenvalue calculation, the highest eigenvalue is 6.27 and highest positive and real eigenvectors are 9.42809042e-02+0.j, 1.16252589e-01-0.32620252j, 1.06776739e-01+0.j belong to web page 5. Therefore, the most important web page is web page 5. 
